internal/integration/rhp/v4/rhp4_test.go

```diff
@@ -0,0 +1,1383 @@
+		t.Fatal(err)
+	} else if err := <-results; err != nil {
+		t.Fatal(err)
+	}
+
+	testutil.MineAndSync(t, hn, w.Address(), int(n.MaturityDelay+20))
+
+	transport := testRenterHostPair(t, hostKey, hn, zap.NewNop())
+
+	settings, err := rhp4.RPCSettings(context.Background(), transport)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fundAndSign := &fundAndSign{w, renterKey}
+	renterAllowance, hostCollateral := types.Siacoins(100), types.Siacoins(200)
+	formResult, err := rhp4.RPCFormContract(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, hostKey.PublicKey(), settings.WalletAddress, proto4.RPCFormContractParams{
+		RenterPublicKey: renterKey.PublicKey(),
+		RenterAddress:   w.Address(),
+		Allowance:       renterAllowance,
+		Collateral:      hostCollateral,
+		ProofHeight:     cm.Tip().Height + 50,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	revision := formResult.Contract
+
+	assertLastRevision := func(t *testing.T) {
+		t.Helper()
+
+		rs, err := rhp4.RPCLatestRevision(context.Background(), transport, revision.ID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		lastRev := rs.Contract
+		if !reflect.DeepEqual(lastRev, revision.Revision) {
+			t.Log(lastRev)
+			t.Log(revision.Revision)
+			t.Fatalf("expected last revision to match")
+		}
+
+		sigHash := cm.TipState().ContractSigHash(revision.Revision)
+		if !renterKey.PublicKey().VerifyHash(sigHash, lastRev.RenterSignature) {
+			t.Fatal("renter signature invalid")
+		} else if !hostKey.PublicKey().VerifyHash(sigHash, lastRev.HostSignature) {
+			t.Fatal("host signature invalid")
+		}
+	}
+	assertLastRevision(t)
+
+	cs := cm.TipState()
+	account := proto4.Account(renterKey.PublicKey())
+
+	accountFundAmount := types.Siacoins(25)
+	fundResult, err := rhp4.RPCFundAccounts(context.Background(), transport, cs, renterKey, revision, []proto4.AccountDeposit{
+		{Account: account, Amount: accountFundAmount},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	revision.Revision = fundResult.Revision
+	assertLastRevision(t)
+
+	token := proto4.AccountToken{
+		Account:    account,
+		ValidUntil: time.Now().Add(time.Hour),
+	}
+	tokenSigHash := token.SigHash()
+	token.Signature = renterKey.SignHash(tokenSigHash)
+
+	// store random sectors
+	roots := make([]types.Hash256, 0, 10)
+	for i := 0; i < 10; i++ {
+		var sector [proto4.SectorSize]byte
+		frand.Read(sector[:])
+		root := proto4.SectorRoot(&sector)
+
+		writeResult, err := rhp4.RPCWriteSector(context.Background(), transport, settings.Prices, token, bytes.NewReader(sector[:]), proto4.SectorSize)
+		if err != nil {
+			t.Fatal(err)
+		} else if writeResult.Root != root {
+			t.Fatal("root mismatch")
+		}
+		roots = append(roots, root)
+	}
+
+	// corrupt a random root
+	excludedIndex := frand.Intn(len(roots))
+	roots[excludedIndex] = frand.Entropy256()
+
+	// append the sectors to the contract
+	appendResult, err := rhp4.RPCAppendSectors(context.Background(), transport, cs, settings.Prices, renterKey, revision, roots)
```

Comment:
Maybe we can have one sector that isn't appended to the contract, mine a bunch of blocks at the end and check that only the pinned sectors remain on the host.

Reply:
Pruning is tested elsewhere. I don't see a good reason to include it here since it's not really part of RHP4.

Reply:
Pruning is tested elsewhere. I don't see a good reason to also test it here.